### PR TITLE
Placeholder

### PR DESCRIPTION
**Pull request by markrwilliams**
Description will be entered later.

## Contributor Checklist:

* [x] There is an associated ticket in Trac. https://twistedmatrix.com/trac/ticket/9647
* [x] The changes pass minimal style checks (see: https://twistedmatrix.com/trac/wiki/TwistedDevelopment#GettingYourPatchAccepted )
* [x] I have created a newsfragment in src/twisted/newsfragments/ 
* [x] I have updated the automated tests.

### CODE REVIEW

File: src/twisted/web/_newclient.py [hunk 1]
```diff
@@ -579,6 +581,58 @@ def connectionLost(self, reason):
...
+
+
+
+def _ensureValidMethod(method):
+    """
+    An HTTP method is an HTTP token, which consists of any visible
+    ASCII character that is not a delimiter (i.e. one of
+    C{"(),/:;<=>?@[\\]{}}.)
+
+    @param method: The method to check
```
Comment (glyph) [inline src/twisted/web/_newclient.py]: `@type`?

File: src/twisted/web/_newclient.py [hunk 2]
```diff
@@ -579,6 +581,58 @@ def connectionLost(self, reason):
...
+    @return: The method if it is valid
```
Comment (glyph) [inline src/twisted/web/_newclient.py]: `@rtype` ?

File: src/twisted/web/_newclient.py [hunk 3]
```diff
@@ -579,6 +581,58 @@ def connectionLost(self, reason):
...
+        return method
+    raise ValueError("Invalid method {!r}".format(method))
+
+
+
+_VALID_URI = re.compile(br'\A[\x21-\x7e]+\Z')
+
+
+
+def _ensureValidURI(uri):
```
Comment (glyph) [inline src/twisted/web/_newclient.py:628]: `@param` `@type` `@rtype`

File: src/twisted/web/_newclient.py [hunk 4]
```diff
@@ -664,8 +718,15 @@ def _writeHeaders(self, transport, TEorCL):
         # method would probably be good.  It would be nice if this method
         # weren't limited to issuing HTTP/1.1 requests.
         requestLines = []
-        requestLines.append(b' '.join([self.method, self.uri,
-            b'HTTP/1.1\r\n']))
+        requestLines.append(
+            b' '.join(
+                [
+                    _ensureValidMethod(self.method),
```
Comment (glyph) [inline src/twisted/web/_newclient.py:740]: augh mutability :(

File: src/twisted/web/client.py [hunk 5]
```diff
@@ -1717,6 +1726,7 @@ def request(self, method, uri, headers=None, bodyProducer=None):
 
         @see: L{twisted.web.iweb.IAgent.request}
         """
+        uri = _ensureValidURI(uri.strip())
```
Comment (glyph) [inline src/twisted/web/client.py:1729]: Should the `.strip()` go in `_ensureValidURI` ?

**Review by glyph — APPROVED** [review]

File: src/twisted/web/client.py (hunk 5 above)
Comment (markrwilliams) [reply src/twisted/web/client.py:1729]: I don't think so, because not all APIs that end up needing this strip their URI.

**Pull request merged**